In [1]:
import sys
!{sys.executable} -m pip install --upgrade pip
!{sys.executable} -m pip install -e ../bayesnet
!{sys.executable} -m pip install matplotlib
!{sys.executable} -m pip install networkx
!{sys.executable} -m pip install graphviz
!{sys.executable} -m pip install datetime
!{sys.executable} -m pip install grpcio
!{sys.executable} -m pip install --no-cache-dir --upgrade numpy
!{sys.executable} -m pip install pomegranate==0.14.0



Obtaining file:///home/opencog/bayesnet
  Preparing metadata (setup.py) ... done
  Attempting uninstall: covid-bayes
    Found existing installation: covid-bayes 1.0
    Can't uninstall 'covid-bayes'. No files were found to uninstall.
  Running setup.py develop for covid-bayes


In [2]:
!{sys.executable} -m pip install -r requirements.txt
!sh buildproto.sh

In [3]:
import sn_bayes
from sn_bayes.utils import any_of
from sn_bayes.utils import all_of
from sn_bayes.utils import avg
from sn_bayes.utils import if_then_else
from sn_bayes.utils import bayesInitialize
from sn_bayes.utils import addCpt
from sn_bayes.utils import dependency
from sn_bayes.utils import non_cpt_descriptions

import sn_service.service_spec.bayesian_pb2
from sn_service.service_spec.bayesian_pb2 import BayesianNetwork




In the independent case, Anyof is correct in the way it treats the variables independently

In [4]:

#EXAMPLE  any_of mitigated without dependency


        bayesianNetwork0 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork0.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        

        discreteDistribution = bayesianNetwork0.discreteDistributions.add()
        discreteDistribution.name = "disorders_of_lipid_metabolism"
        variable = discreteDistribution.variables.add()
        variable.name = "disorders_of_lipid_metabolism"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_disorders_of_lipid_metabolism"
        variable.probability = 0.50
        

        discreteDistribution = bayesianNetwork0.discreteDistributions.add()
        discreteDistribution.name = "diabetes"
        variable = discreteDistribution.variables.add()
        variable.name = "diabetes"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_diabetes"
        variable.probability = 0.50
        

                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork0,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        
        

        cpt["metabolic_syndrome"] = all_of(bayesianNetwork0,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )

        

        outstr = outstr + non_cpt_descriptions(bayesianNetwork0)
        outstr = outstr + addCpt(bayesianNetwork0,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity0 = bayesInitialize(bayesianNetwork0)
        longevity0.bake()
        bayesianNetwork0


discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
discreteDistributions {
  name: "disorders_of_lipid_metabolism"
  variables {
    name: "disorders_of_lipid_metabolism"
    probability: 0.5
  }
  variables {
    name: "no_disorders_of_lipid_metabolism"
    probability: 0.5
  }
}
discreteDistributions {
  name: "diabetes"
  variables {
    name: "diabetes"
    probability: 0.5
  }
  variables {
    name: "no_diabetes"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "metabolic_disease"
  randomVariables {
    name: "hypertension"
  }
  randomVariables {
    name: "disorders_of_lipid_metabolism"
  }
  randomVariables {
    name: "diabetes"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    randomVariableValues {
      n

In [5]:
findsum = longevity0.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125])

In [6]:
findsum.sum()

1.0000000000000002

In [7]:
 longevity0.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.5,
            "no_disorders_of_lipid_metabolism" : 0.5
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.5,
            "no_diabetes" : 0.5
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.2500000000000001,
            "metabolic_disease" : 0.7499999999999999
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 1.0,
            "metabol

In [8]:
 longevity0.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.5,
            "no_disorders_of_lipid_metabolism" : 0.5
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.5,
            "no_diabetes" : 0.5
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 0.7499999999999999,
            "metabolic_syndrome" : 0.2

### Problem 1:  correlation vs. cause in the net

Lets view A as a test of B and C, and B as a test of C
I want to estimate my risk of condition C 
lets also simplify that the only way you can get B is from A and the only way you can get C is from B

scenario 1

Say we have A leads to B half the time and B leads to C half the time.
then A leads to C a forth of the time.
Say A occurs naturally at 50%
Then B would occur at 25%
and C would occur at 12.5%



C:
so A causes C at sensitivity .25 and specificity .99
and B causes C at sensitivity .5 and specificity .99

B:
A causes B at sensitivity .5 and specificity .99

so how is it that I can get the same answer for C, when it looks like this:
C:
and B causes C at sensitivity .5 and specificity .99

### Problem 2:  mitigation of anyof and all of.  Anyof treats the variables as independent.  Is this made up for by dependencies before the anyof CPT?  

for anyof, chance of true = abc(.125) + abNOTc (.125) + aNOTbNOTc (.25) = .5 ; chance of false = NOTaNOTbNOTc = .5

for allof, chance of true = abc = .125 ; chance of false = .875


In [9]:
##### EXAMPLE  any_of mitigated without dependency


        bayesianNetwork1 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork1.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork1,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":0.5, "specificity":0.999999})],
        {"disorders_of_lipid_metabolism":0.25,"no_disorders_of_lipid_metabolism":0.75}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork1,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork1,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":0.25, "specificity":0.99999}),
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"sensitivity":0.5, "specificity":0.99999})],
        {"diabetes":0.125,"no_diabetes":0.875}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork1,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        

        cpt["metabolic_syndrome"] = all_of(bayesianNetwork1,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )

        outstr = outstr + non_cpt_descriptions(bayesianNetwork1)
        outstr = outstr + addCpt(bayesianNetwork1,cpt) 

        from sn_bayes.utils import bayesInitialize
        longevity1 = bayesInitialize(bayesianNetwork1)
        longevity1.bake()
        bayesianNetwork1

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.9999940000359998}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 5.9999640001664886e-06}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.14285726530622742}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.8571427346937726}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0

/home/opencog/bayesnet/sn_bayes/utils.py:1991: OptimizeWarning: A_eq does not appear to be of full row rank. To improve performance, check the problem formulation for redundant equality constraints.
  opt = linprog(c=obj, A_ub=lhs_ineq, b_ub=rhs_ineq,A_eq=lhs_eq, b_eq=rhs_eq,  bounds=bnd,method="revised simplex")


discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.5
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.5
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "no

In [10]:
findsum = longevity1.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.125, 0.125, 0.   , 0.25 , 0.   , 0.   , 0.   , 0.5  ])

In [11]:
findsum.sum()

1.0

In [12]:
 longevity1.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.0,
            "no_disorders_of_lipid_metabolism" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.0,
            "no_diabetes" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 1.0,
            "metabolic_disease" : 0.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 1.0,
            "metabolic_syndrome" : 0.0
        }
 

In [13]:
 longevity1.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.5,
            "no_disorders_of_lipid_metabolism" : 0.5000000000000001
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.25000000000000033,
            "no_diabetes" : 0.7499999999999998
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 0.8749999999

In [14]:
#EXAMPLE  any_of mitigated without dependency


        bayesianNetwork1_1 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork1_1.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork1_1,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":0.5, "specificity":0.999999})],
        {"disorders_of_lipid_metabolism":0.25,"no_disorders_of_lipid_metabolism":0.75}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork1_1,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork1_1,cpt,
        [
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"sensitivity":0.5, "specificity":0.99999})],
        {"diabetes":0.125,"no_diabetes":0.875}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork1_1,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )


        cpt["metabolic_syndrome"] = all_of(bayesianNetwork1_1,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )
        outstr = outstr + non_cpt_descriptions(bayesianNetwork1_1)
        outstr = outstr + addCpt(bayesianNetwork1_1,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity1_1 = bayesInitialize(bayesianNetwork1_1)
        longevity1_1.bake()
        bayesianNetwork1_1
                

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.9999940000359998}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 5.9999640001664886e-06}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.14285726530622742}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.8571427346937726}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.5
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.5
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "no

In [15]:
findsum = longevity1_1.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.125, 0.125, 0.   , 0.25 , 0.   , 0.   , 0.   , 0.5  ])

In [16]:
findsum.sum()

1.0

In [17]:
 longevity1_1.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.0,
            "no_disorders_of_lipid_metabolism" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.0,
            "no_diabetes" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 1.0,
            "metabolic_disease" : 0.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 1.0,
            "metabolic_syndrome" : 0.0
        }
 

In [18]:
 longevity1_1.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.5000000000000001,
            "no_disorders_of_lipid_metabolism" : 0.5
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.2500000000000004,
            "no_diabetes" : 0.7499999999999997
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 0.87499999999

    Using Relative risk instead of sensitivity and specificity, we know that the relative risk of getting B if you have A vs if you do not is, 50% vs 0%, or infinity.  The relative risk of getting C if you have A is 25% vs 0% or infinity


In [19]:

        bayesianNetwork2 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork2.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork2,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":9999})],
        {"disorders_of_lipid_metabolism":0.25,"no_disorders_of_lipid_metabolism":0.75}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork2,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork2,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":9999}),
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"relative_risk":9999})],
        {"diabetes":0.125,"no_diabetes":0.875}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork2,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        

        cpt["metabolic_syndrome"] = all_of(bayesianNetwork2,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )

        outstr = outstr + non_cpt_descriptions(bayesianNetwork2)
        outstr = outstr + addCpt(bayesianNetwork2,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity2 = bayesInitialize(bayesianNetwork2)
        longevity2.bake()
        bayesianNetwork2
                

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.49995}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.50005}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 4.999999999999449e-05}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.99995}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ]), array([0. , 0. , 0.5, 0.5])]


opt
     con: array([ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
       -5.55111512e-17,  1.11022302e-16])
     fun: 4.0
 message: 'Optimization terminated successfully.'
     nit: 9
   slack: array([ 8.75000000e-01,  1.00000000e+00,  6.25000012e-01,  4.99999988e-01,
        1.25000000e-01, -0.00000000e+00,  3.74999988e-01,  5.00000012e-01,
        8.75050000e-01,  9.99950000e-01,  3.74950005e-01,  7.50049995e-01,
        1.24950000e-01,  4.99999984e-05,  6.25049995e-01,  2.49950005e-01])
  status: 0
 success: True
       x: array([0.      , 0.      , 1.      , 0.249925, 1.      , 1.      ,
       0.      , 0.750075])
window
0.5
obj
[1. 1. 1. 1. 1. 1. 1. 1.]
lhs_eq
[array([1., 0., 0., 0., 1., 0., 0., 0.]), array([0., 1., 0., 0., 0., 1., 0., 0.]), array([0., 0., 1., 0., 0., 0., 1., 0.]), array([0., 0., 0., 1., 0., 0., 0., 1.]), array([2.49950005e-01, 2.50050007e-01, 4.99999984e-05, 4.99949988e-01,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.4999000132083893
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.5001000165939331
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 9.999999747378752e-05
  }
  conditionalProbabilityRows {
    randomVariableValues {
      nam

In [20]:
findsum = longevity2.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([1.25000002e-01, 1.24950004e-01, 0.00000000e+00, 2.50050008e-01,
       0.00000000e+00, 4.99999987e-05, 0.00000000e+00, 4.99949992e-01])

In [21]:
findsum.sum()

1.0000000066029315

In [22]:
 longevity2.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 9.999999913432131e-05,
            "no_disorders_of_lipid_metabolism" : 0.9999000000008657
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.0,
            "no_diabetes" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.9999000000008658,
            "metabolic_disease" : 9.999999913432126e-05
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_s

In [23]:
 longevity2.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.49989999831020837,
            "no_disorders_of_lipid_metabolism" : 0.5001000016897916
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.24999999745024265,
            "no_diabetes" : 0.7500000025497573
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome

In [24]:

        bayesianNetwork2_1 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork2_1.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork2_1,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":9999})],
        {"disorders_of_lipid_metabolism":0.25,"no_disorders_of_lipid_metabolism":0.75}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork2_1,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork2_1,cpt,
        [
           # ({"hypertension":["hypertension"]},{"relative_risk":9999}),
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"relative_risk":9999})],
        {"diabetes":0.125,"no_diabetes":0.875}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork2_1,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        

        cpt["metabolic_syndrome"] = all_of(bayesianNetwork2_1,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )

        outstr = outstr + non_cpt_descriptions(bayesianNetwork2_1)
        outstr = outstr + addCpt(bayesianNetwork2_1,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity2_1 = bayesInitialize(bayesianNetwork2_1)
        longevity2_1.bake()
        bayesianNetwork2_1

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.49995}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.50005}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 4.999999999999449e-05}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.99995}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ]), array([0. , 0. , 0.5, 0.5])]


opt
     con: array([4.4408921e-16, 0.0000000e+00, 0.0000000e+00, 0.0000000e+00])
     fun: 1.9999999999999996
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([ 0.50004999,  0.87485002,  0.25      ,  0.875     , -0.        ,
        0.125     ,  0.25004999,  0.12485003])
  status: 0
 success: True
       x: array([0.49999999, 0.        , 0.50000001, 1.        ])
window
0.25
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.25, 0.75, 0.  , 0.  ]), array([0.  , 0.  , 0.25, 0.75])]
rhs_eq
[1, 1, 0.125, 0.875]
lhs_ineq
[array([0.  , 0.75, 0.  , 0.  ]), array([0.25, 0.  , 0.  , 0.  ]), array([0.  , 0.  , 0.  , 0.75]), array([0.  , 0.  , 0.25, 0.  ]), array([-0.  , -0.75, -0.  , -0.  ]), array([-0.25, -0.  , -0.  , -0.  ]), array([-0.  , -0.  , -0.  , -0.75]), array([-0.  , -0.  , -0.25, -0.  ])]
rhs_ineq
[0.25004999000100975, 0.7498500200965343, 1.0, 0.7501499799034655, -0.0, -0.2498500200965344, -0.7499500099989902, -0.250149

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.4999000132083893
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.5001000165939331
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 9.999999747378752e-05
  }
  conditionalProbabilityRows {
    randomVariableValues {
      nam

In [25]:
findsum = longevity2_1.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([1.24825053e-01, 1.25124953e-01, 5.00016496e-05, 2.50000005e-01,
       2.49700033e-05, 2.50299954e-05, 9.99732992e-05, 4.99850015e-01])

In [26]:
findsum.sum()

1.0000000013642418

In [27]:
 longevity2_1.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.4999000000564382,
            "no_disorders_of_lipid_metabolism" : 0.5000999999435618
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.24975010294192654,
            "no_diabetes" : 0.7502498970580735
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome"

In [28]:
 longevity2_1.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 9.999999983274342e-05,
            "no_disorders_of_lipid_metabolism" : 0.9999000000001673
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.0002498866108810005,
            "no_diabetes" : 0.999750113389119
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.9996501383779475,
            "metabolic_disease" : 0.0003498616220524757
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
     


lets say when a is present b is present half the time, and that when a is absent b is present a forth of the time.  
since a naturally occurs half the time, so b naturally occurs 3/8 of the time, .375

the rr for a would be .5/.25 = 2

sensitivity of test a for b would be .5

specificity of test a for b would be .25




lets say when b is present c is present half the time, and that when b is absent c is present a forth of the time. 
since b naturally occurs 3/8 of the time, c naturally occurs  .375 x .5 + 1-.375 * .25 = .34375

rr for b would be .5/.25 = 2

sensitivity of test b for c would be .5

specificity of test b for c would be .25



the risk of c in the absense of a is 

a absent b present .5*.25 = .125  risk of c:.5

a present b present .5*.5 = .25  risk of c:.5

a absent b absent .5*.75 = .375 risk of c:.25

a present b absent .5*.5 = .25  risk of c:.5



risk of c in the presence of a is (.25 * .5 + .25 * .5)*2= .5

risk of c in the absense of a is (.125 * .5 + .375 * .25)*2= .3125

rr for a on c = .5/.3125 = 1.6

sensitivity of test a for c would be .5

specificity of test a for c would be .3125


anyof = 1-~a(.5)~b(.75)~c(.75) = .71875

allof =a(.5)b(.5)c(.5) =  .125

In [29]:
        bayesianNetwork3 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork3.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork3,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":2})],
        {"disorders_of_lipid_metabolism":0.375,"no_disorders_of_lipid_metabolism":0.625}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork3,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork3,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":1.6}),
           # ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"relative_risk":2})
        ],
        {"diabetes":0.34375,"no_diabetes":0.65625}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork3,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        


        cpt["metabolic_syndrome"] = all_of(bayesianNetwork3,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )
        outstr = outstr + non_cpt_descriptions(bayesianNetwork3)
        outstr = outstr + addCpt(bayesianNetwork3,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity3 = bayesInitialize(bayesianNetwork3)
        longevity3.bake()
        bayesianNetwork3
                

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.5}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.5}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.25}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.75}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ]), array([0. , 0. , 0.5, 0.5])]
rhs_eq
[1, 1, 0.375, 0.625]


opt
     con: array([-0.3125 ,  0.     ,  0.     , -0.15625])
     fun: 2.3125
 message: 'The problem appears infeasible, as the phase one auxiliary problem terminated successfully with a residual of 1.6e-01, greater than the tolerance 1e-12 required for the solution to be considered feasible. Consider increasing the tolerance to be greater than 1.6e-01. If this tolerance is unnaceptably large, the problem is likely infeasible.'
     nit: 2
   slack: array([0.5    , 0.34375, 0.5    , 0.5    , 0.     , 0.15625, 0.     ,
       0.     ])
  status: 2
 success: False
       x: array([0.65865385, 0.02884615, 0.65384615, 0.97115385])
window
0.375
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ]), array([0. , 0. , 0.5, 0.5])]
rhs_eq
[1, 1, 0.34375, 0.65625]
lhs_ineq
[array([0. , 0.5, 0. , 0. ]), array([0.5, 0. , 0. , 0. ]), array([0. , 0. , 0. , 0.5]), array([0. , 0. , 0.5, 0. ]), array([-0. , -0.5, -0. , -0. ]), array([-0.5, -0. , -0. , -

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.625
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.375
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.125
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVaria

In [30]:
findsum = longevity3.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.12770433, 0.18479567, 0.0766226 , 0.1108774 , 0.01742788,
       0.04507212, 0.12199519, 0.31550481])

In [31]:
findsum.sum()

1.0

In [32]:
 longevity3.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.12500000000000047,
            "no_disorders_of_lipid_metabolism" : 0.8749999999999994
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.27884614467620883,
            "no_diabetes" : 0.7211538553237912
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.6310096234083171,
            "metabolic_disease" : 0.3689903765916829
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
 

In [33]:
 longevity3.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.6249999999999999,
            "no_disorders_of_lipid_metabolism" : 0.37500000000000006
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.4086538553237916,
            "no_diabetes" : 0.5913461446762083
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome"

In [34]:

        bayesianNetwork3_1 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork3_1.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork3_1,cpt,
        [
            ({"hypertension":["hypertension"]},{"relative_risk":2})],
        {"disorders_of_lipid_metabolism":0.375,"no_disorders_of_lipid_metabolism":0.625}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork3_1,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork3_1,cpt,
        [
           
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"relative_risk":2})],
        {"diabetes":0.34375,"no_diabetes":0.65625}
        )
        
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork3_1,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        


        cpt["metabolic_syndrome"] = all_of(bayesianNetwork3_1,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )
        outstr = outstr + non_cpt_descriptions(bayesianNetwork3_1)
        outstr = outstr + addCpt(bayesianNetwork3_1,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity3_1 = bayesInitialize(bayesianNetwork3_1)
        longevity3_1.bake()
        bayesianNetwork3_1

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.5}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.5}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.25}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.75}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ]), array([0. , 0. , 0.5, 0.5])]
rhs_eq
[1, 1, 0.375, 0.625]


[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.375, 0.625, 0.   , 0.   ]), array([0.   , 0.   , 0.375, 0.625])]
rhs_eq
[1, 1, 0.34375, 0.65625]
lhs_ineq
[array([0.   , 0.625, 0.   , 0.   ]), array([0.375, 0.   , 0.   , 0.   ]), array([0.   , 0.   , 0.   , 0.625]), array([0.   , 0.   , 0.375, 0.   ]), array([-0.   , -0.625, -0.   , -0.   ]), array([-0.375, -0.   , -0.   , -0.   ]), array([-0.   , -0.   , -0.   , -0.625]), array([-0.   , -0.   , -0.375, -0.   ])]
rhs_ineq
[0.5625, 0.8125, 1.0, 0.8125000000000001, -0.0, -0.18749999999999994, -0.4375, -0.1875000000000001]
bnd
[(0, 1.0), (0, 1.0), (0, 1.0), (0, 1.0)]
opt
     con: array([2.22044605e-16, 1.11022302e-16, 0.00000000e+00, 0.00000000e+00])
     fun: 1.9999999999999998
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([0.40625, 0.625  , 0.53125, 0.625  , 0.15625, 0.     , 0.03125,
       0.     ])
  status: 0
 success: True
       x: array([0.5 , 0.25, 0.5 , 0.75])
window
0.28125
cpt_rows


discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.625
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.375
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.125
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVaria

In [35]:
findsum = longevity3_1.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.15625 , 0.15625 , 0.046875, 0.140625, 0.03125 , 0.03125 ,
       0.109375, 0.328125])

In [36]:
findsum.sum()

1.0

In [37]:
 longevity3_1.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.1250000000000007,
            "no_disorders_of_lipid_metabolism" : 0.8749999999999993
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.2812500000000004,
            "no_diabetes" : 0.7187499999999997
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.6289062499999993,
            "metabolic_disease" : 0.3710937500000008
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
   

In [38]:
 longevity3_1.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.625,
            "no_disorders_of_lipid_metabolism" : 0.3750000000000001
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.4062500000000001,
            "no_diabetes" : 0.5937499999999998
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome" : 0.746093749

In [39]:
        bayesianNetwork4 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork4.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        
        
        
        
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork4,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":.5,"specificity":.25})],
        {"disorders_of_lipid_metabolism":0.375,"no_disorders_of_lipid_metabolism":0.625}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork4,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork4,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":.5,"specificity":.3125}),
           # ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"sensitivity":.5,"specificity":.25})
        ],
        {"diabetes":0.34375,"no_diabetes":0.65625}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork4,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        

        cpt["metabolic_syndrome"] = all_of(bayesianNetwork4,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )

        outstr = outstr + non_cpt_descriptions(bayesianNetwork4)
        outstr = outstr + addCpt(bayesianNetwork4,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity4 = bayesInitialize(bayesianNetwork4)
        longevity4.bake()
        longevity4.predict_proba({}) 
        bayesianNetwork4
                

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.2857142857142857}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.7142857142857143}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.5454545454545454}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.45454545454545453}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ])

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.034090910106897354
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.9659090638160706
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.7159090638160706
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name

In [40]:
findsum = longevity4.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.00256675, 0.0144787 , 0.0727247 , 0.41022983, 0.19219191,
       0.16576263, 0.07626663, 0.06577882])

In [41]:
findsum.sum()

0.9999999720603229

In [42]:
 longevity4.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.7159090851518236,
            "no_disorders_of_lipid_metabolism" : 0.2840909148481762
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.5369170904159546,
            "no_diabetes" : 0.4630829095840455
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.13155764743428688,
            "metabolic_disease" : 0.8684423525657132
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
  

In [43]:
 longevity4.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.034090910995887756,
            "no_disorders_of_lipid_metabolism" : 0.9659090890041123
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.15058290958404585,
            "no_diabetes" : 0.8494170904159543
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrom

In [44]:

        bayesianNetwork4_1 = BayesianNetwork()
        cpt = {}
        outstr = '' 



        discreteDistribution = bayesianNetwork4_1.discreteDistributions.add()
        discreteDistribution.name = "hypertension"
        variable = discreteDistribution.variables.add()
        variable.name = "hypertension"
        variable.probability = 0.50
        variable = discreteDistribution.variables.add()
        variable.name = "no_hypertension"
        variable.probability = 0.50
        
        
 
 
        cpt["disorders_of_lipid_metabolism"] = dependency(bayesianNetwork4_1,cpt,
        [
            ({"hypertension":["hypertension"]},{"sensitivity":.5,"specificity":.25})],
        {"disorders_of_lipid_metabolism":0.375,"no_disorders_of_lipid_metabolism":0.625}
        )
        
     

        outstr = outstr + addCpt(bayesianNetwork4_1,cpt) 
        cpt = {}

        cpt["diabetes"] = dependency(bayesianNetwork4_1,cpt,
        [
            #({"hypertension":["hypertension"]},{"sensitivity":.5,"specificity":.3125}),
            ({"disorders_of_lipid_metabolism":["disorders_of_lipid_metabolism"]},{"sensitivity":.5,"specificity":.25})
        ],
        {"diabetes":0.34375,"no_diabetes":0.65625}
        )
        
                 
        cpt["metabolic_disease"] = any_of(bayesianNetwork4_1,cpt,
                {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_disease","no_metabolic_disease"]
                 
         )
        


        cpt["metabolic_syndrome"] = all_of(bayesianNetwork4_1,cpt,
       {
         'hypertension':{ "hypertension"},
         "disorders_of_lipid_metabolism":{"disorders_of_lipid_metabolism"},
         "diabetes":{"diabetes"}
         },
         ["metabolic_syndrome","no_metabolic_syndrome"]
                 
         )
        outstr = outstr + non_cpt_descriptions(bayesianNetwork4_1)
        outstr = outstr + addCpt(bayesianNetwork4_1,cpt) 
        from sn_bayes.utils import bayesInitialize
        longevity4_1 = bayesInitialize(bayesianNetwork4_1)
        longevity4_1.bake()
        longevity4_1.predict_proba({}) 
        bayesianNetwork4_1

start timing...
vdict
{'hypertension': ['hypertension', 'no_hypertension']}
vlist
[['hypertension', 'no_hypertension']]
i:c
0
('hypertension',)
c[pos[k]]
hypertension
i:c
1
('no_hypertension',)
c[pos[k]]
no_hypertension
rhs_inequality_equation2
{'hypertension': {'hypertension': 0.2857142857142857}}
lhs_inequality_equation2
{'hypertension': {'hypertension': array([0.5, 0. , 0. , 0. ])}}
rhs_inequality_equation4
{'hypertension': {'hypertension': 0.7142857142857143}}
lhs_inequality_equation4
{'hypertension': {'hypertension': array([0. , 0. , 0.5, 0. ])}}
rhs_inequality_equation1
{'hypertension': {'hypertension': 0.5454545454545454}}
lhs_inequality_equation1
{'hypertension': {'hypertension': array([0. , 0.5, 0. , 0. ])}}
rhs_inequality_equation3
{'hypertension': {'hypertension': 0.45454545454545453}}
lhs_inequality_equation3
{'hypertension': {'hypertension': array([0. , 0. , 0. , 0.5])}}
obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.5, 0.5, 0. , 0. ])

obj
[1. 1. 1. 1.]
lhs_eq
[array([1., 0., 1., 0.]), array([0., 1., 0., 1.]), array([0.375, 0.625, 0.   , 0.   ]), array([0.   , 0.   , 0.375, 0.625])]
rhs_eq
[1, 1, 0.34375, 0.65625]
lhs_ineq
[array([0.   , 0.625, 0.   , 0.   ]), array([0.375, 0.   , 0.   , 0.   ]), array([0.   , 0.   , 0.   , 0.625]), array([0.   , 0.   , 0.375, 0.   ]), array([-0.   , -0.625, -0.   , -0.   ]), array([-0.375, -0.   , -0.   , -0.   ]), array([-0.   , -0.   , -0.   , -0.625]), array([-0.   , -0.   , -0.375, -0.   ])]
rhs_ineq
[0.8241279069767442, 0.5713235294117647, 0.8008720930232558, 1.0, -0.1991279069767442, -0.0, -0.1758720930232558, -0.4286764705882353]
bnd
[(0, 1.0), (0, 1.0), (0, 1.0), (0, 1.0)]
opt
     con: array([1.     , 1.     , 0.34375, 0.65625])
     fun: 0.0
 message: 'The problem is (trivially) infeasible because a singleton row in the upper bound constraints is inconsistent with the bounds.'
     nit: 0
   slack: array([ 0.82412791,  0.57132353,  0.80087209,  1.        , -0.19912791,
   

discreteDistributions {
  name: "hypertension"
  variables {
    name: "hypertension"
    probability: 0.5
  }
  variables {
    name: "no_hypertension"
    probability: 0.5
  }
}
conditionalProbabilityTables {
  name: "disorders_of_lipid_metabolism"
  randomVariables {
    name: "hypertension"
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.034090910106897354
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "hypertension"
    }
    randomVariableValues {
      name: "no_disorders_of_lipid_metabolism"
    }
    probability: 0.9659090638160706
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name: "no_hypertension"
    }
    randomVariableValues {
      name: "disorders_of_lipid_metabolism"
    }
    probability: 0.7159090638160706
  }
  conditionalProbabilityRows {
    randomVariableValues {
      name

In [45]:
findsum = longevity4_1.probability([
         ['hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "metabolic_syndrome"], ['hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "disorders_of_lipid_metabolism",
         "no_diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "diabetes",
    "metabolic_disease",
    "no_metabolic_syndrome"], ['no_hypertension',
         "no_disorders_of_lipid_metabolism",
         "no_diabetes",
    "no_metabolic_disease",
    "no_metabolic_syndrome"],
])
findsum

array([0.00040107, 0.01664439, 0.25880681, 0.22414772, 0.00842246,
       0.34953207, 0.07611965, 0.0659258 ])

In [46]:
findsum.sum()

0.9999999699648471

In [47]:
 longevity4_1.predict_proba({"hypertension":"no_hypertension"}) 

array(['no_hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.715909084015331,
            "no_disorders_of_lipid_metabolism" : 0.28409091598466896
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.1690842234596642,
            "no_diabetes" : 0.8309157765403357
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.2360556240634565,
            "metabolic_disease" : 0.7639443759365435
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
   

In [48]:
 longevity4_1.predict_proba({"hypertension":"hypertension"}) 

array(['hypertension',
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "disorders_of_lipid_metabolism" : 0.034090910811884395,
            "no_disorders_of_lipid_metabolism" : 0.9659090891881157
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "diabetes" : 0.5184157750982867,
            "no_diabetes" : 0.4815842249017133
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_disease" : 0.0,
            "metabolic_disease" : 1.0
        }
    ],
    "frozen" : false
},
       {
    "class" : "Distribution",
    "dtype" : "str",
    "name" : "DiscreteDistribution",
    "parameters" : [
        {
            "no_metabolic_syndrome